### PR TITLE
Generate production Csv file

### PR DESCRIPTION
**Pull request by IsaackMwenda**
@as2388  Have added Docker Scrips, production file export & main pipeline module.

### CODE REVIEW

File: docker-run.sh [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        esac
+    done 
+
+#Check that the correct number of arguments were provided
+if [[ $# -ne 17 ]]; then 
+    echo "Usage: ./docker-run.sh 
+    [--profile-cpu <profile-output-path>] 
+    [--drive-upload <drive-auth-file> <messages-drive-path> <individuals-drive-path> <production-drive-path>]
+    <user> <phone-number-uuid-table-path> 
+    <internet_working_poll-input-path> <water_filter_poll-input-path> <waste_disposal-input-path> <demog-input-path> <prev-coded-dir>
```
Comment (as2388) [inline docker-run.sh]: please be consistent with the '-'s vs '_'s here.

I believe hyphens are standard in unix tool argument names

File: docker-run.sh [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+fi
+
+CMD="
+    $GSUTIL_CP_CMD \
+
+    pipenv run $PROFILE_CPU_CMD python -u test_pipeline.py $DRIVE_UPLOAD_ARG \
+    \"$USER\" /data/phone-number-uuid-table-input.json \
+    /data/internet_working-input.json /data/water_filter-input.json /data/waste_disposal-input.json \
+    /data/demog-input.json /data/prev-coded \ /data/output.json /data/ouput-icr /data/coded
```
Comment (as2388) [inline docker-run.sh]: There is an extraneous backslash here

Comment (as2388) [reply docker-run.sh]: And I think you need one at the end of the line, I'm surprised this actually works

File: docker-run.sh [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+docker cp "$container:/data/output-icr/." "$OUTPUT_ICR_DIR"
+
+mkdir -p "$OUTPUT_CODED_DIR"
+docker cp "$container:/data/output-messages-csv" "$OUTPUT_MESSAGES_CSV"
+
+mkdir -p "$(dirname "$OUTPUT_MESSAGES_CSV")"
+docker cp "$container:/data/output-messages.csv" "$OUTPUT_MESSAGES_CSV"
+
+mkdir -p "$(dirname "$OUTPUT_INDIVIDUALS_CSV")"
+docker cp "$container:/data/output-individuals.csv" "$OUTPUT_INDIVIDUALS_CSV"
```
Comment (as2388) [inline docker-run.sh]: I'm confused here as well. As far as I can tell, this file isn't being produced, so shouldn't this line be failing?

File: project_test/combine_raw_datasets.py [hunk 4]
```diff
@@ -1,15 +1,15 @@
...
     def combine_raw_datasets(user, shows_datasets, survey_datasets):
         data = []
 
         for show_dataset in shows_datasets:
             data.extend(show_dataset)
 
-        for surveys_datasets in survey_datasets:
-            TracedData.update_iterable(user, "avf_phone_id", data, survey_dataset, "survey_responses")
+        for survey_dataset in survey_datasets:
+            TracedData.update_iterable(user, "avf_phone_id", data, survey_datasets, "survey_responses")
```
Comment (as2388) [inline project_test/combine_raw_datasets.py]: survey_dataset?

File: project_test/lib/__init__.py [hunk 5]
```diff
@@ -0,0 +1,3 @@
+#from .icr_tools import ICRTools
+from .message_filters import MessageFilters
+from .pipeline_configurations import PipelineConfiguration
```
Comment (as2388) [inline project_test/lib/__init__.py]: No new line at end of file.

There will be a switch in your editor that forces files to end with a new line.

File: project_test/lib/message_filters.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+                    continue
+        return non_empty
+
+    @staticmethod
+    def filter_time_range(messages, time_key, starts_time, end_time):
+        return [td for td in messages if starts_time <= isoparse(td.get(time_key)) <= end_time]
+
+    def filter_noise(messages,messages_key, noise_fn):
+        return [td for td in messages if not noise_fn(td.get(messages_key))]
+
```
Comment (as2388) [inline project_test/lib/message_filters.py]: ditto

File: project_test/production_file.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    def generate(data, production_csv_output_path):
+        production_keys = ["uid"]
+        for plan in PipelineConfiguration.TEST_SHOWS_CODING_PLAN:
+            if plan.raw_field not in production_keys:
+                production_keys.append(plan.raw_field)
+        for plan in PipelineConfiguration.SURVEY_CODING_PLANS:
+            if plan.raw_field not in production_keys:
+                production_keys.append(plan.raw_field)
+                
+        #Not perfoming any message filtering on this test-pipeline
```
Comment (as2388) [inline project_test/production_file.py]: In that case do we need message_filters.py?

File: test_pipeline.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    production_csv_output_path = args.production_csv_output_path
+
+    shows_paths = [water_filter_input_path, internet_working_input_path, waste_disposal_input_path]
+
+    # Load phone number <-> UID Table
+    print("Loading phone Number <-> UUID Table...")
+    with open(phone_number_uuid_table_path, "r") as f:
+        phone_number_uuid_table_path = PhoneNumberUuidTable.load(f)
+
+    #Load Messages
```
Comment (as2388) [inline test_pipeline.py]: Convention in Python is to put a space after the '#'. Python editors should be able to point this out to you automatically, you may need to flick a switch somewhere to enable "PEP-8" support

**Review by as2388 — CHANGES_REQUESTED** [review]

**Comment by IsaackMwenda** [comment]
@as2388  Please take another look

**Comment by IsaackMwenda** [comment]
@as2388  Please take another look. Thanks

File: code_schemes/internet_working.json [hunk 9]
```diff
@@ -0,0 +1,91 @@
+{
+    "Version": "0.0.0.1",
+    "Name": "Internet Working Diagnosis?",
```
Comment (as2388) [inline code_schemes/internet_working.json]: Drop the '?'

File: project_test/lib/message_filters.py (hunk 6 above)
Comment (as2388) [reply project_test/lib/message_filters.py]: still no ne

File: project_test/production_file.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
+from core_data_modules.traced_data.io import TracedDataCSVIO
+
+from project_test.lib import PipelineConfiguration 
+
+class ProductionFile(object):
+    @staticmethod
+    def generate(data, production_csv_output_path):
+        production_keys = ["uid"]
+        for plan in PipelineConfiguration.TEST_SHOWS_CODING_PLAN:
```
Comment (as2388) [inline project_test/production_file.py]: Prefer TEST_SHOWS_CODING_PLANS if it is a list

File: test_pipeline.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+        production_csv_drive_file_name = os.path.basename(production_csv_drive_path)
+        drive_client_wrapper.update_or_create(production_csv_output_path, production_csv_drive_dir,
+                                            target_file_name=production_csv_drive_file_name,
+                                            target_folder_is_shared=True)
+        print("Files successfully uploaded")
+
+    else:
+        print("Not uploading to Google Drive")
+    
+    print("Writing TracedData to file....")
```
Comment (as2388) [inline test_pipeline.py]: I know I did it in this order on ADSS, but I am starting to think we should write the TracedData before the Drive upload. The problem with the order used here is that if writing the TracedData fails, then we have analysis files in Drive but no TracedData associated with them...

**Review by as2388 — APPROVED** [review]

File: test_pipeline.py (hunk 11 above)
Comment (IsaackMwenda) [reply test_pipeline.py]: That makes sense thank @as2388

**Comment by IsaackMwenda** [comment]
Hey @lukechurch PTAL!

**Comment by lukechurch** [comment]
Hi @IsaackMwenda - I'm not going to have much time to review this in the next couple of days. As long as @as2388 says it's ok, that's enough for me on this one!

**Review by lukechurch — APPROVED** [review]

**Pull request merged**